### PR TITLE
feat(source): Support AWS privatelink for CREATE TABLE command

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Support AWS PrivateLink for the CREATE TABLE command that uses a cloud-hosted Kafka source (MSK).

```sql
CREATE TABLE mytable with (
  connector = 'kafka',
  topic = 'kafka_1_partition_topic',
  properties.bootstrap.server = '127.0.0.1:29092',
  connection.name = 'pl1',
  privatelink.targets = '[{"port": 8001}, {"port": 8002}]'
) ROW FORMAT JSON;
```

### <samp>🤖 Generated by Copilot at 7088ca4</samp>

This pull request refactors and simplifies the code for handling Kafka connector connections in various handlers and catalog modules. It introduces a common utility function `resolve_connection_in_with_option` to get and validate connection ids from with options. It also improves the reference counting logic of connections in the `CatalogManager` by using two helper functions from `utils.rs`.

related: #8771 

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Connector (sources & sinks)
- SQL commands, functions, and operators

### Release note
Support AWS PrivateLink for the CREATE TABLE command that uses a cloud-hosted Kafka source 

</details>

### CODE REVIEW

**Comment by StrikeW** [comment]
> Is it a user-facing change? Besides, what's its difference from our former CREATE CONNECTION command?

CREATE CONNECTION is only for creating a privatelink connection, then the connection can be used by source, sink, table that needs to access a MSK service in another vpc.

File: src/frontend/src/utils/with_options.rs [hunk 1]
```diff
@@ -103,6 +110,39 @@ impl WithOptions {
...
 
+#[inline(always)]
+fn is_kafka_connector(with_options: &WithOptions) -> bool {
+    let Some(connector) = with_options.inner().get(UPSTREAM_SOURCE_KEY).map(|s| s.to_lowercase()) else {
+        return false;
+    };
+    connector == KAFKA_CONNECTOR
+}
+
+pub(crate) fn resolve_connection_in_with_option(
```
Comment (WillyKidd) [inline src/frontend/src/utils/with_options.rs:121]: maybe we should rename it to `resolve_connection_id_with_option` 🤣

**Review by WillyKidd — APPROVED** [review]
Rest LGTM :)

File: src/frontend/src/utils/with_options.rs (hunk 1 above)
Comment (WillyKidd) [reply src/frontend/src/utils/with_options.rs:121]: oh I see what you mean 😅😅

**Comment by CharlieSYH** [comment]
Documented at https://www.risingwave.dev/docs/upcoming/create-source-kafka/#create-source-with-aws-privatelink-connection.

**Pull request merged**